File: 算法基础课(修订版)/循环/P14076/verify_acm.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Solution {
    public int[] solve(int[] arr, int l, int r) {
        l--;
        r--;
        int maxValue = arr[l];
        List<Integer> maxIndices = new ArrayList<>();
        maxIndices.add(l + 1);
        for (int i = l + 1; i <= r; ++i) {
            if (arr[i] > maxValue) {
                maxValue = arr[i];
                maxIndices.clear();
                maxIndices.add(i + 1);
            } else if (arr[i] == maxValue) {
                maxIndices.add(i + 1);
            }
        }
        int[] result = new int[1 + maxIndices.size()];
        result[0] = maxValue;
        for (int i = 0; i < maxIndices.size(); ++i) {
            result[i + 1] = maxIndices.get(i);
        }
        return result;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int n = scanner.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; ++i) {
            arr[i] = scanner.nextInt();
        }

        int q = scanner.nextInt();
        Solution solution = new Solution();
        while (q-- > 0) {
            int l = scanner.nextInt();
            int r = scanner.nextInt();
            int[] result = solution.solve(arr, l, r);

            System.out.println(result[0]);
            for (int i = 1; i < result.length; ++i) {
                System.out.print(result[i]);
                if (i + 1 < result.length) {
                    System.out.print(" ");
                }
            }
            System.out.println();
        }

        scanner.close();
    }
}

File: 算法基础课(修订版)/循环/P14076/verify_acm.cc
#include <iostream>
#include <vector>
#include <utility>
using namespace std;

class Solution {
public:
    pair<int, vector<int>> solve(vector<int>& arr, int l, int r) {
        l--;
        r--;
        int maxValue = arr[l];
        vector<int> maxIndices = {l + 1};
        for (int i = l + 1; i <= r; ++i) {
            if (arr[i] > maxValue) {
                maxValue = arr[i];
                maxIndices.clear();
                maxIndices.push_back(i + 1);
            } else if (arr[i] == maxValue) {
                maxIndices.push_back(i + 1);
            }
        }
        return {maxValue, maxIndices};
    }
};

int main() {
    int n;
    cin >> n;
    vector<int> arr(n);
    for (int i = 0; i < n; ++i) {
        cin >> arr[i];
    }

    int q;
    cin >> q;

    Solution solution;
    while (q--) {
        int l, r;
        cin >> l >> r;
        auto result = solution.solve(arr, l, r);
        cout << result.first << endl;
        for (int i = 0; i < (int)result.second.size(); ++i) {
            cout << result.second[i] << (i + 1 == (int)result.second.size() ? "" : " ");
        }
        cout << endl;
    }

    return 0;
}
